testing/src/test_node.rs

```diff
@@ -0,0 +1,310 @@
+    /// Starts the usual periodic activities of a test node.
+    pub fn run_periodic_tasks(&self) {
+        self.send_pings();
+        self.update_peers();
+    }
+}
+
+/// Automated handshake handling for the test nodes.
+#[async_trait::async_trait]
+impl Handshake for TestNode {
+    async fn perform_handshake(&self, mut conn: Connection) -> io::Result<Connection> {
+        // Guard against double (two-sided) connections.
+        let mut locked_peers = self.state.peers.lock().await;
+
+        if locked_peers.contains(&conn.addr) {
+            return Err(io::ErrorKind::AlreadyExists.into());
+        }
+
+        let own_addr = self.node().listening_addr()?;
+        let peer_addr = conn.addr;
+
+        let genesis_block_header = Testnet2::genesis_block().header();
+
+        // Send a challenge request to the peer.
+        let own_request = SnarkosMessage::ChallengeRequest(MESSAGE_VERSION, own_addr.port(), 0, 0);
+        trace!(parent: self.node().span(), "sending a challenge request to {}", peer_addr);
+        let msg = own_request.serialize().unwrap();
+        let len = u32::to_le_bytes(msg.len() as u32);
+        conn.writer().write_all(&len).await?;
+        conn.writer().write_all(&msg).await?;
+
+        let mut buf = [0u8; 1024];
+
+        // Read the challenge request from the peer.
+        conn.reader().read_exact(&mut buf[..4]).await?;
+        let len = u32::from_le_bytes(buf[..4].try_into().unwrap()) as usize;
+        conn.reader().read_exact(&mut buf[..len]).await?;
+        let peer_request = SnarkosMessage::deserialize(&buf[..len]);
+
+        // Register peer's listening address.
+        let peer_listening_addr =
+            if let Ok(Message::ChallengeRequest(peer_version, peer_listening_port, _nonce, _block_height)) = peer_request {
+                if peer_version < MESSAGE_VERSION {
+                    warn!(parent: self.node().span(), "dropping {} due to outdated version ({})", peer_addr, peer_version);
+                    return Err(io::ErrorKind::InvalidData.into());
+                }
+                trace!(parent: self.node().span(), "received a challenge request from {}", peer_addr);
+                SocketAddr::from((peer_addr.ip(), peer_listening_port))
+            } else {
+                error!(parent: self.node().span(), "invalid challenge request from {}", peer_addr);
+                return Err(io::ErrorKind::InvalidData.into());
+            };
+
+        // Respond with own challenge request.
+        let own_response = SnarkosMessage::ChallengeResponse(genesis_block_header.clone());
+        trace!(parent: self.node().span(), "sending a challenge response to {}", peer_addr);
+        let msg = own_response.serialize().unwrap();
+        let len = u32::to_le_bytes(msg.len() as u32);
+        conn.writer().write_all(&len).await?;
+        conn.writer().write_all(&msg).await?;
+
+        // Wait for the challenge response to come in.
+        conn.reader().read_exact(&mut buf[..4]).await?;
```

Comment:
Might be good to have these lengths codefied as shared constants?

Reply:
fair; I was just drafting it quickly :+1: